### ISSUE TITLE
works slow

### ISSUE THREAD

**Issue by 532910** [Drive4ik/simple-tab-groups#490]
Switching between groups takes about 5 seconds.
383 tabs total in 19 groups

v4.3.2,  ff 70.0, debian sid

**Comment by grahamperrin** [comment]
#379 led to improved _startup_ times. 

@532910 please review _switching_ times with current versions of Simple Tab Groups and Firefox. 

On a circa 2013 computer with FreeBSD `14.0-CURRENT`: until a few hours ago I had more than 1,400 tabs across many groups in release candidate 3 of Firefox `86.0`. Switching times were no problem.

**Comment by nekohayo** [comment]
It takes between 3 and 5 seconds on my computer (from 2009, running Fedora) to switch between tab groups; I have over 2480 tabs spread across roughly 70 groups.

I use the "discard" option though, so it only keeps the tabs of the current group in RAM at any given time, so the performance "cost" of switching is simply loading one of the tabs of the newly selected group.

3 to 5 seconds to switch groups is frankly not shocking to me.  Yes, I like performance as much as anybody else, but 3-5 secs sounds acceptable considering that number seems to stay constant even with someone like me who has ten times more tabs, so if it isn't a problem that scales with the number of tabs, wel... I'm not the maintainer/developer, but unless you can point out specifically some optimization that can be done in the code to bring it down to less than a second, I would recommend closing this ticket because "it takes about 5 seconds to switch between groups" isn't really actionable as a problem description (nothing tells me it's not a Firefox limitation per se).

**Comment by ZelphirKaltstahl** [comment]
> 3 to 5 seconds to switch groups is frankly not shocking to me. Yes, I like performance as much as anybody else, but 3-5 secs sounds acceptable considering that number seems to stay constant even with someone like me who has ten times more tabs, so if it isn't a problem that scales with the number of tabs, well... I'm not the maintainer/developer, but unless you can point out specifically some optimization that can be done in the code to bring it down to less than a second, I would recommend closing this ticket because "it takes about 5 seconds to switch between groups" isn't really actionable as a problem description (nothing tells me it's not a Firefox limitation per se).

I just browsed the open issues here for the first time thinking: "It is nice, but man, those switchting times are annoying!"

I think there needs to be a good reason for anything as simple as displaying a list of tabs requiring 3-5s in a modern browser. The question is, what is the cause for such a brutal slowdown. Is it some kind of Firefox extension API limitation? What could possibly take 3-5s, when getting a list of tabs and filtering it? In that time one could probably filter a million tabs or more, considering modern CPUs or even parallel processing.

In short: I really like this extension for its functionality, but the delay when switchting groups is _not_ acceptable without good reasoning behind it. This delay makes me consider, whether I could do better by simply using `%` in the URL bar, instead of sorting things in groups, to find my tabs.

**Comment by ZelphirKaltstahl** [comment]
May I also suggest a title change, to keep this separated from other slowness issues?

"Switching tab groups is slow" as a suggestion.

**Comment by nekohayo** [comment]
For switching, there might be a performance bug in either STG or in Firefox itself. For example, I'm pretty sure there is a [performance bug in Firefox itself when it comes to trying to load all tabs at once](https://bugzilla.mozilla.org/show_bug.cgi?id=1756368), which is what happens when you try to restore STG tabs from a backup.

On startup, there is some slowness too, and I'm not sure if it's Firefox's fault or STG; but that might influence switching times too, potentially.

On opening the popup to choose a group to switch to (not the same thing as the actual switching), there is a slightly annoying 2-3 seconds delay when you have many tab groups like I do, that's issue #1012.
